### ISSUE TITLE
When making changes to scratch/index.ts during development, then the new version is not loaded in Scratch unless you Ctrl-F5 hard refresh

### ISSUE THREAD

**Issue by vorburger** [vorburger/minecraft-storeys-maker#97]
as the title says, this is a big PITA and has caused me massive confusion :-( more than once now: When making changes to scratch/index.ts during development, then the new version is not loaded in Scratch unless you Ctrl-F5 hard refresh

@edewit FYI & let's sort this out somehow...

**Comment by vorburger** [comment]
I ran into this a few days ago, and AGAIN just now during manual testing of #96.

So the scenario here is I didn't use `npm start`, just our `./gradle build` (in root for everything). The ` scratch/dist/minecraft.scratchx.js` is fine, of course; the `scratch*.jar!/static/minecraft.scratchx.js` therefore obviously too. At first I suspected the Flash based Scratch container that loads the extension to be too stupid to figure out it had to reload. But then I noticed... even http://localhost:7070/minecraft.scratchx.js opened in a browser tab had the old version, not including a change I just made. So figured what seemed to be happening is that an HTTP 304 _Not Modified_. And indeed after I hit Ctr-F5 to force a reload, like back when we did web dev in the 90s :smiling_imp: it showed the new version. Retrying `/make` and reloading the Flash based Scratch container then worked.

What's the root cause of that? I have a suspicion that it is my `ch.vorburger.minecraft.storeys.web.StaticWebServerVerticle` which is using `io.vertx.ext.web.handler.StaticHandler` which may be at fault here...

**Comment by vorburger** [comment]
I can reproduce this, but it's important NOT to manually open http://localhost:7070/minecraft.scratchx.js in  browser first, but to directly `/make` and open Scratch. The Chrome Developer Tools Network shows that it DOES get a 200 and not 304 (I think that was a misinterpretation on my part, I initially a 304 when opening the direct URL on a tab AFTER opening Scratch; that seems normal, actually).

However, the file's content when looked at directly inside Chrome indeed is the old one! And Chrome shows **(from disk cache)** in _Size_, which doesn't seem right. The `scratch_ext.js` has this:

    var loadingURL;
    lib.loadExternalJS = function (url) {
        var scr = document.createElement("script");
        scr.src = url;// + "?ts=" + new Date().getTime();
        loadingURL = url;
        document.getElementsByTagName("head")[0].appendChild(scr);
    };

Digging through io.vertx.ext.web.handler.impl.StaticHandlerImpl, I've found `setFilesReadOnly(false)`, but in testing with adding that to `StaticWebServerVerticle` it doesn't seem to help. That makes some sense, as looking at the related code it seems to be related to a MEMORY cache, so after rebuild and Minecraft server restart, that couldn't cache anything from before the change.

So it's Chrome that doesn't "reload" and uses an old version from disk, despite having gotten a 200. This is probably because it sees a `last-modified` which does seem of, from before the change (unless I'm misinterpreting due to TZ diff) ...

**Comment by vorburger** [comment]
> adding that (`setFilesReadOnly(false)`) to StaticWebServerVerticle it doesn't seem to help

or does it... it seems to work now, with it - AFTER I did a one time clear cache. I think that's because Vert.x's filesReadOnly default true made it ALWAYS send a 302, so that Disk Cache still had that, from before the change? Not 100% sure, and doesn't entirely add up to completely make sense to me, but seems to work now; so pushed as f67094dc993e12f31536583abc5272d111f1aa05.

PS: There is also `setCachingEnabled(false)` and I've toyed with that as well, it will make it NEVER cache and ALWAYS resend the JS. That seems more and too extreme - we don't want/need that; and it does now seem to work with just `.setFilesReadOnly(false)` without needing `setCachingEnabled(false)`.

PPS: In dev, using `npm start` (and `-Dstoreys_jsURL=http://localhost:3000/minecraft.scratchx.js`) also works around this. But given what I learnt about, I now believe this actually wasn't even only a dev env issue, but a cache problem which may affected our production as well; not sure, but better safe than sorry.

**Comment by vorburger** [comment]
This problem hit @edewit and again tody while hacking together at DevConf.US in Boston.

Hopefully https://github.com/vorburger/minecraft-storeys-maker/commit/33bf22b8c4805bd4963c9a5746ed36553ad6dce8 fixes this for good. I tested this locally: "Simulating" with a `minecraft-storeys-maker/web/src/main/resources/static/hello.txt` it even without this change after a restart, the new content did get served on an F5. But with this `setCachingEnabled(false)` it now updated even live/hot (while running, when changed on classpath, it re-reads it from the ClassLoader every time) and -more importantly- when I opened a new tab to http://localhost:9090/hello.txt WITHOUT F5, both with and without restart, it would should the new content - which is what we want.

This is slightly less efficient in production now, but for the time we keep wasting on this in dev, that's fine.
